### PR TITLE
Bug fix for non-zero position ids not working for full deepseek model

### PR DESCRIPTION
**Pull request by barci2**
### Problem description
Non-zero position ids were not working for full deepseek model. This was because the reference model did not have the cache properly initialized, but instead the input was padded for positions lower than the actual input position id. This was equivalent to zero-initializing the cache for the decoder and mla, since the linear maps for MLA didn't have any bias, but wasn't fine for the full model, where the 0-index token does not have a zero representation. 

### What's changed
- Added initialization functions for both reference and ttnn model mla caches.
- The tests now initialize the caches randomly instead of zeroed-out
- Extracted out the context-shortened version of hf_config into a pytest fixture
- Simplified test code for MLA, decoder block and model
- Removed is_padding_layer from decoder block and added model config sharing between submodules (since it should be copyable and serializable, so it shouldn't keep any runtime mutable state)
- Small code cleanup for MLA
- Changed the way softmax scale is decided for MLA to match the HF reference implementation
- Removed reference_forwards.py in favor of a general function for running reference HF models.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17582858069) CI passes
- [ ] [Deepseek](https://github.com/tenstorrent/tt-metal/actions/runs/17590989048) CI passes

### CODE REVIEW

File: models/demos/deepseek_v3/utils/test_utils.py [hunk 1]
```diff
@@ -61,73 +91,229 @@ def add_inv_scale_to_state_dict(
...
             output_state_dict[name] = tensor
             continue
+        assert tensor.ndim == len(block_shape), "Weight tensors must have the same dimensionality as the block shape"
 
-        tensor_quant, dequant_scale = quantize_fp8_blockwise(tensor, block_shape)
-        output_state_dict[name] = tensor_quant
-        output_state_dict[name + "_scale_inv"] = dequant_scale
+        scale_tensor = get_quant_scale(tensor, block_shape)
+        output_state_dict[name] = dequantize(tensor, scale_tensor, block_shape).to(torch.float8_e4m3fn)
+        output_state_dict[name + "_scale_inv"] = (1.0 / scale_tensor).float()
```
Comment (kpaigwar) [inline models/demos/deepseek_v3/utils/test_utils.py]: With this new quantization scheme,
Please verify, MoE decoder block test also passes.

File: models/demos/deepseek_v3/tt/model_1d.py [hunk 2]
```diff
@@ -207,72 +193,92 @@ def transfer_row(
...
-            The tensor after transferring the row
+            dst_row_idx: Destination row index
+            mesh_shape: Mesh shape
+            topology: Topology
+            semaphore: Semaphore object
         """
-
-        mesh_shape = cfg["mesh_shape"]
+        if src_row_idx == dst_row_idx:
+            return x
```
Comment (kpaigwar) [inline models/demos/deepseek_v3/tt/model_1d.py]: Great! I also hit upon this case.

File: models/demos/deepseek_v3/tt/model_1d.py [hunk 3]
```diff
@@ -43,70 +39,62 @@ def convert_weights(
...
+        assert mesh_rows > 0, "Number of rows in a mesh is greater than zero."
+        assert num_layers >= 0, "Number of layers cannot be negative."
 
-        if num_rows <= 0:
-            raise ValueError("Number of rows must be greater than zero.")
-        if num_layers < 0:
-            raise ValueError("Number of layers cannot be negative.")
+        num_meta_layers = cls.get_num_meta_layers(num_layers, mesh_rows)
+        layer_indices = torch.arange(num_meta_layers * mesh_rows).reshape(num_meta_layers, mesh_rows)
+        return layer_indices >= num_layers, layer_indices
```
Comment (kpaigwar) [inline models/demos/deepseek_v3/tt/model_1d.py]: Since, you're already refactoring this part.
Can you extend this function to get meta layer mapping from a start_layer_num=x rather assuming start layer is always 0.
Usecase : for getting metalayer mapping for MoE decoder blocks which begins from layer_idx=3

**Review by kpaigwar — CHANGES_REQUESTED** [review]
overall looks good, just make sure all tests passes locally or on CI

File: models/demos/deepseek_v3/reference/modeling_deepseek.py [hunk 4]
```diff
@@ -1150,15 +1152,17 @@ class DeepseekV3PreTrainedModel(PreTrainedModel):
     _supports_cache_class = True
 
     def _init_weights(self, module):
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/reference/modeling_deepseek.py]: nit: remove this if not used now or add a comment why are we removing this code but keeping the function as a place holder?

File: models/demos/deepseek_v3/tests/test_mla_1d.py [hunk 5]
```diff
@@ -86,187 +66,157 @@ def get_cache_on_host(tt_cache: ttnn.Tensor, row_idx: int, mesh_device: ttnn.Mes
 )
 @pytest.mark.parametrize(
     "module_path",
-    ["model.layers.0.self_attn"],
+    [None, "model.layers.0.self_attn"],
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tests/test_mla_1d.py:71]: module_path can be a simple variable in test_forward_pass() and parametrize weights_type to ["random", "real"], to check for random or real weights. That will keep the code more readable.

File: models/demos/deepseek_v3/tests/test_mla_1d.py [hunk 6]
```diff
@@ -86,187 +66,157 @@ def get_cache_on_host(tt_cache: ttnn.Tensor, row_idx: int, mesh_device: ttnn.Mes
...
     if mode == "decode":
-        # TT Shape: [1, seq_len, batch_size, hidden_size]
-        torch_input = torch_input.permute(1, 0, 2)
+        torch_input = torch_input.permute(1, 0, 2)  # [1, seq_len, batch_size, hidden_size]
 
     tt_input = ttnn.from_torch(
         torch_input.unsqueeze(0),
         device=mesh_device,
-        mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, dims=(None, -1), mesh_shape=mesh_shape),
+        mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, mesh_device.shape, (None, -1)),
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tests/test_mla_1d.py]: nit: keep arg_name = value. It would help to keep code more readable.

File: models/demos/deepseek_v3/tests/test_mla_1d.py [hunk 7]
```diff
@@ -86,187 +66,157 @@ def get_cache_on_host(tt_cache: ttnn.Tensor, row_idx: int, mesh_device: ttnn.Mes
...
-
-    if mode == "decode":
-        position_idxs_tensor = ttnn.from_torch(
-            torch.tensor(position_idxs),
+    position_ids_tensor = (
+        ttnn.from_torch(
+            position_ids,
             device=mesh_device,
-            mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, dims=(None, 0), mesh_shape=mesh_shape),
+            mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, mesh_device.shape, (None, 0)),
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tests/test_mla_1d.py]: ditto as line#154

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 5 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_mla_1d.py:71]: use the same structure of  "weights_type" in other tests too,

File: models/demos/deepseek_v3/tests/test_model.py [hunk 8]
```diff
@@ -65,102 +49,65 @@ def test_forward_pass(
...
-    logger.info("Running Torch reference forward pass")
+    # Get reference IO
+    logger.info("Setting up reference IO")
     if module_path is None:
-        reference_output = reference_forward(
-            reference_model,
-            torch_input,
-            position_ids=position_idxs,
-            mode=mode,
+        reference_model = DeepseekV3Model(hf_config_short).eval().to(torch.bfloat16)
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tests/test_model.py]: nit: why don't you keep load_reference_model() ?

File: models/demos/deepseek_v3/tt/decoder_block/moe_decoder_block.py [hunk 9]
```diff
@@ -89,17 +82,10 @@ def decode_mlp_config(
...
-        assert mesh_device.shape[0] == len(
-            is_padding_layer
-        ), "Number of mesh device rows must match the number of padding or non-padding layers"
         return {
             "shared_expert": SharedExpert.decode_model_config(hf_config, mesh_device),
-            "moe": [
-                None if is_padding else MoE.decode_model_config(hf_config, mesh_device)
-                for is_padding in is_padding_layer
-            ],
+            "moe": [MoE.decode_model_config(hf_config, mesh_device)] * mesh_device.shape[0],
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tt/decoder_block/moe_decoder_block.py]: Why is this change needed? Wouldn't we have some padding layers with MOEs?

File: models/demos/deepseek_v3/tt/model_1d.py [hunk 10]
```diff
@@ -207,72 +193,92 @@ def transfer_row(
...
-            x = cls.transfer_row(x, row_idx, (row_idx + 1) % cls.NUM_MLP_ROWS, cfg)
+            cls.transfer_row(x, row_idx, (row_idx + 1) % cfg["num_rows"], **cfg["transfer_row"])
+
+        x_resharded = ttnn.to_memory_config(x, **cfg["norm_reshard"])
+        # ttnn.deallocate(x)
 
-        x = ttnn.to_memory_config(x, **cfg["norm_reshard"])
-        x = DistributedRMSNorm.forward_decode(x, cfg["norm"])
+        x_norm = DistributedRMSNorm.forward_decode(x_resharded, cfg["norm"])
+        # ttnn.deallocate(x_resharded)
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tt/model_1d.py]: uncomment this to deallocate x_resharded!

**Review by pprajapatiTT — CHANGES_REQUESTED** [review]

File: models/demos/deepseek_v3/tt/model_1d.py [hunk 11]
```diff
@@ -207,72 +193,92 @@ def transfer_row(
...
+        return x_norm
 
     @classmethod
     def forward_prefill(
         cls,
         x: ttnn.Tensor,
         cfg: RunPrefillConfig,
     ) -> ttnn.Tensor:
-        return x
+        raise NotImplementedError("Prefill mode is not implemented.")
```
Comment (pprajapatiTT) [inline models/demos/deepseek_v3/tt/model_1d.py:292]: This is nice..

File: models/demos/deepseek_v3/tt/model_1d.py (hunk 3 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/model_1d.py]: I don't think this is necessary, since you can just add a number to the tensor on the caller's end.

File: models/demos/deepseek_v3/utils/test_utils.py (hunk 1 above)
Comment (barci2) [reply models/demos/deepseek_v3/utils/test_utils.py]: Yeah I always run the whole deepseek CI before each PR

File: models/demos/deepseek_v3/reference/modeling_deepseek.py (hunk 4 above)
Comment (barci2) [reply models/demos/deepseek_v3/reference/modeling_deepseek.py]: idk I just wanted to keep this in case we do need that functionality back, since we are ultimately changing the reference code

File: models/demos/deepseek_v3/tests/test_model.py (hunk 8 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_model.py]: to make the code shorter - I don't think we need a whole function just to instantiate that single class. And if we do want functions for those, let's make it a general framework for all the tests somehow (though I'm not sure it would add any value in terms of abstraction)

File: models/demos/deepseek_v3/reference/modeling_deepseek.py (hunk 4 above)
Comment (barci2) [reply models/demos/deepseek_v3/reference/modeling_deepseek.py]: Actually, I just realized we don't need to uncomment this

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 5 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_mla_1d.py:71]: I kind of thought that the module path is meaningless in terms of real weights. Also, we soon are going to be removing "real" weights IO generation entirely and using the pre-cached reference IO for model testing, so I think that this kind of structure does not duplicate the information. Do you have a strong preference here? We could call the weights_type param different, but then again module_path does not make any sense in the context of random weights

**Review by yieldthought — APPROVED** [review]

**Review by kpaigwar — APPROVED** [review]

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 5 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_mla_1d.py:71]: It's better to keep same naming conventions in all the tests. For now we can keep "weights_type", when we move to new pre-cached ref io, we can change all the tests and keep same new name in all of the tests, you can choose any name then.

File: models/demos/deepseek_v3/tests/test_model.py (hunk 8 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_model.py]: When we attach lm_head we need DeepseekV3ForCausalLM as reference model, so keeping a function with param with type of ref model needed would be helpful. So i think we should keep the func. 
It's nit. 

But may be when i add lm head i can put back the func. Your choice for now.

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 5 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_mla_1d.py:71]: yes but this naming convention is not used in any other tests. MLP uses this convention, so do RMSNorm, MLA, DecoderBlock and Model tests.

File: models/demos/deepseek_v3/tests/test_model.py (hunk 8 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_model.py]: let's put that function in later then

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 5 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_mla_1d.py:71]: That is why I changed it, to keep the naming convention consistent

Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_mla_1d.py:71]: yeah, we need to start using clear naming for random vs real weights. approving the PR for now but i would love to see this fixed.

**Review by pprajapatiTT — APPROVED** [review]

**Comment by barci2** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/17582831857
